File: Task 2/Java/app/src/main/java/dev/ogabek/java/network/VolleyHttp.java
package dev.ogabek.java.network;

import androidx.annotation.Nullable;

import com.android.volley.Request;
import com.android.volley.toolbox.StringRequest;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

import dev.ogabek.java.MyApplication;
import dev.ogabek.java.model.Employee;

public class VolleyHttp {

    private static final String TAG = VolleyHttp.class.toString();

    public static final boolean IS_TESTER = false;
    public static final String SERVER_DEVELOPMENT = "http://dummy.restapiexample.com/api/v1/employees";
    public static final String SERVER_PRODUCTION = "https://62220b53666291106a1b35d3.mockapi.io/";

    public static String server(String url) {
        if (IS_TESTER) {
            return SERVER_DEVELOPMENT + url;
        } else {
            return SERVER_PRODUCTION + url;
        }
    }

    public static HashMap<String, String> header() {
        HashMap<String, String> header = new HashMap<>();
        header.put("Content-type", "application/json; charset=UTF-8");
        return header;
    }

    // Request Methods

    public static void get(String api, HashMap<String, String> params, VolleyHandler volleyHandler) {
        StringRequest request = new StringRequest(
                Request.Method.GET, server(api),
                volleyHandler::onSuccess,
                error -> volleyHandler.onError(error.getLocalizedMessage())
        ) {
            @Nullable
            @Override
            protected Map<String, String> getParams() {
                return params;
            }
        };
        MyApplication.instance.addToRequestQueue(request);
    }

    public static void post(String api, HashMap<String, String> body, VolleyHandler volleyHandler) {
        StringRequest request = new StringRequest(
                Request.Method.POST, server(api),
                volleyHandler::onSuccess,
                error -> volleyHandler.onError(error.getLocalizedMessage())
        ) {
            @Override
            public Map<String, String> getHeaders() {
                return header();
            }

            @Override
            public byte[] getBody() {
                return new JSONObject((Map<String, String>) body).toString().getBytes();
            }
        };
        MyApplication.instance.addToRequestQueue(request);
    }

    public static void put(String api, HashMap<String, String> body, VolleyHandler volleyHandler) {
        StringRequest request = new StringRequest(
                Request.Method.PUT, server(api),
                volleyHandler::onSuccess,
                error -> volleyHandler.onError(error.getLocalizedMessage())
        ) {
            @Override
            public Map<String, String> getHeaders() {
                return header();
            }

            @Override
            public byte[] getBody() {
                return new JSONObject((Map<String, String>) body).toString().getBytes();
            }
        };
        MyApplication.instance.addToRequestQueue(request);
    }

    public static void delete(String api, VolleyHandler volleyHandler) {
        StringRequest request = new StringRequest(
                Request.Method.DELETE, server(api),
                volleyHandler::onSuccess,
                error -> volleyHandler.onError(error.getLocalizedMessage())
        );
        MyApplication.instance.addToRequestQueue(request);
    }

    public static final String API_GET_ALL = "m6-l7-t2";
    public static final String API_GET_SINGLE = "m6-l7-t2/";
    public static final String API_PUT = "m6-l7-t2/";
    public static final String API_DELETE = "m6-l7-t2/";
    public static final String API_POST = "m6-l7-t2";

    // Params

    public static HashMap<String, String> paramsEmpty() {
        return new HashMap<>();
    }

    public static HashMap<String, String> paramsPost(Employee employee) {
        HashMap<String, String> body = new HashMap<>();
        body.put("employee_name", employee.getEmployee_name());
        body.put("employee_salary", employee.getEmployee_salary());
        body.put("employee_age", employee.getEmployee_age());
        body.put("profile_image", employee.getProfile_image());
        return body;
    }

    public static HashMap<String, String> paramsPut(Employee employee) {
        HashMap<String, String> body = new HashMap<>();
        body.put("employee_name", employee.getEmployee_name());
        body.put("employee_salary", employee.getEmployee_salary());
        body.put("employee_age", employee.getEmployee_age());
        body.put("profile_image", employee.getProfile_image());
        return body;
    }

}

File: Task 2/Kotlin/app/src/main/java/dev/ogabek/kotlin/network/VolleyHttp.kt
package dev.ogabek.kotlin.network

import com.android.volley.Request
import com.android.volley.toolbox.StringRequest
import dev.ogabek.kotlin.MyApplication
import dev.ogabek.kotlin.model.Employee
import org.json.JSONObject

object VolleyHttp {
    private val TAG = VolleyHttp::class.java.toString()
    const val IS_TESTER = false
    private const val SERVER_DEVELOPMENT = "http://dummy.restapiexample.com/api/v1/employees"
    private const val SERVER_PRODUCTION = "https://62220b53666291106a1b35d3.mockapi.io/"
    fun server(url: String): String {
        return if (IS_TESTER) {
            SERVER_DEVELOPMENT + url
        } else {
            SERVER_PRODUCTION + url
        }
    }

    fun header(): HashMap<String, String> {
        val header = HashMap<String, String>()
        header["Content-type"] = "application/json; charset=UTF-8"
        return header
    }

    // Request Methods
    operator fun get(api: String, params: HashMap<String, String>, volleyHandler: VolleyHandler) {
        val request: StringRequest = object : StringRequest(
            Method.GET, server(api),
            volleyHandler::onSuccess,
            { error -> volleyHandler.onError(error.localizedMessage) }
        ) {
            override fun getParams(): MutableMap<String, String> {
                return params
            }
        }
        MyApplication.instance!!.addToRequestQueue(request)
    }

    fun post(api: String, body: HashMap<String, String>, volleyHandler: VolleyHandler) {
        val request: StringRequest = object : StringRequest(
            Method.POST, server(api),
            volleyHandler::onSuccess,
            { error -> volleyHandler.onError(error.localizedMessage) }
        ) {
            override fun getHeaders(): MutableMap<String, String> {
                return header();
            }

            override fun getBody(): ByteArray {
                return JSONObject(body as HashMap<*, *>).toString().toByteArray()
            }
        }
        MyApplication.instance!!.addToRequestQueue(request)
    }

    fun put(api: String, body: HashMap<String, String>, volleyHandler: VolleyHandler) {
        val request: StringRequest = object : StringRequest(
            Request.Method.PUT, server(api),
            volleyHandler::onSuccess,
            { error -> volleyHandler.onError(error.localizedMessage!!+ error.networkResponse) }
        ) {
            override fun getHeaders(): MutableMap<String, String> {
                return header();
            }

            override fun getBody(): ByteArray {
                return JSONObject(body as HashMap<*, *>).toString().toByteArray()
            }
        }
        MyApplication.instance!!.addToRequestQueue(request)
    }

    fun delete(api: String, volleyHandler: VolleyHandler) {
        val request = StringRequest(
            Request.Method.DELETE, server(api),
            volleyHandler::onSuccess
        ) { error -> volleyHandler.onError(error.localizedMessage) }
        MyApplication.instance!!.addToRequestQueue(request)
    }

    const val API_GET_ALL = "m6-l7-t2"
    const val API_GET_SINGLE = "m6-l7-t2/"
    const val API_PUT = "m6-l7-t2/"
    const val API_DELETE = "m6-l7-t2/"
    const val API_POST = "m6-l7-t2"

    // Params
    fun paramsEmpty(): HashMap<String, String> {
        return HashMap()
    }

    fun paramsPost(employee: Employee): HashMap<String, String> {
        val body = HashMap<String, String>()
        body["employee_name"] = employee.employee_name
        body["employee_salary"] = employee.employee_salary
        body["employee_age"] = employee.getEmployee_age()
        body["profile_image"] = employee.profile_image
        return body
    }

    fun paramsPut(employee: Employee): HashMap<String, String> {
        val body = HashMap<String, String>()
        body["employee_name"] = employee.employee_name
        body["employee_salary"] = employee.employee_salary
        body["employee_age"] = employee.getEmployee_age()
        body["profile_image"] = employee.profile_image
        return body
    }
}